pkg/aws/ec2: Explicitly set network card index 0

41ks opened the pull request:
A bug was detected in AWS’s ENI attachment API. The validation of available ENI attachment slots fails on instances with multiple network cards when the network card index is not set explicitly. AWS is working on a fix that is expected to be released by January 30, 2026. In the meantime, this change resolves the issue on the Cilium side.

# Example

If an instance with 2 network cards is used and interfaces are attached to the following slots:
- Card Index 0, Device Index 0
- Card Index 1, Device Index 1

This means that the slot (Card Index 0, Device Index 1) is available. But if I try to attach an ENI to it without specifying the network card index and rely on the parameter's default value of 0, I get this error:
```
An error occurred (InvalidParameterValue) when calling the AttachNetworkInterface operation: Instance '<instance-id>' already has an interface attached at device index '1'.
```
If I then retry by explicitly setting Network Card Index to 0, the API call works and the ENI is attached as expected.

# Side note

Cilium only supports the use of one NIC (the primary NIC) on AWS instances. This change does not introduce any new limits on network attachment slots.

## Release note
```release-note
AWS EC2: Fix ENI attachment on multi-network card instances with high-performance networking (EFA) setups
```

liyihuang commented:
maybe I dont understand the basic here.

> DeviceIndex
The position of the network interface in the attachment order. A primary network interface has a device index of 0.

that's like a slot on the physical machine right?

> NetworkCardIndex (request), NetworkCardIndex (response)
The index of the network card. Some instance types support multiple network cards. The primary network interface must be assigned to network card index 0. The default is network card index 0.

Is this the index of the ENI?

If so, why there is device index 0 and card index 1?

In my test environment,

I see the following. I'm pretty confused, do you mind to help me understand it better?

<img width="932" height="516" alt="image" src="https://github.com/user-attachments/assets/5d6019b2-e9f3-4ba1-a918-451cbdd27408" />

41ks committed:
pkg/aws/ec2: Explicitly set network card index 0

A bug was detected in AWS’s ENI attachment API. The validation of available ENI attachment slots fails on instances with multiple network cards when the network card index is not set explicitly. AWS is working on a fix that is expected to be released by January 30. In the meantime, this change resolves the issue on the Cilium side.

Signed-off-by: Alex Melhem <alex.melhem@datadoghq.com>

liyihuang reviewed (approved):
LGTM

julianwiedmann commented:
:wave: could you please describe the user-visible change in the `release-note`? What type of operation was broken before, and works now?
